From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 18 Jul 2026 13:48:30 -0500
Subject: [PATCH] Fish Handle Enderpearl Teleport

This patch sends the enderpearl teleportation to the end of the tick.
Similar to Entity Portal Teleportation, we do not need the custom logic, as worlds still have a main thread,
  we can still rely on the main thread to handle this logic safely.

diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 8d3c6abd387acc4cec27a33b96bcd24255fd01ac..cfe3844b029b050b8677a27672428ac248caf933 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -108,6 +108,37 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = player.getX(), preTeleportY = player.getY(), preTeleportZ = player.getZ();
                         final float preTeleportYRot = player.getYRot(), preTeleportXRot = player.getXRot();
+                        // Fish start - Handle Enderpearl Teleport
+                        if (me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled && !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(owner)) {
+                            // It is possible to use a Folia-like strategy, although their implementation doesn't trigger some events
+                            // We can take advantage of the still-existing main thread to handle this at the end of the tick
+                            // For vanilla-parity reasons, this will only trigger if the ender-pearl and the owner are in different worlds
+                            me.biquaternions.fish.async.AsyncWorldTicking.scheduleVoidForEndOfTick(() -> {
+                                // Pearls disappear with the player on disconnection
+                                ServerPlayer newOwner = player.teleport(new TeleportTransition(
+                                    level, teleportPos, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING).withCause(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
+                                if (newOwner == null) {
+                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                                    return;
+                                }
+                                if (this.random.nextFloat() < 0.05F && level.isSpawningMonsters() && level.getLevelData().getDifficulty() != Difficulty.PEACEFUL) {
+                                    Endermite endermite = EntityTypes.ENDERMITE.create(level, EntitySpawnReason.TRIGGERED);
+                                    if (endermite != null) {
+                                        endermite.snapTo(preTeleportX, preTeleportY, preTeleportZ, preTeleportYRot, preTeleportXRot); // Paper - spawn endermite at pre teleport position as teleport has been moved up
+                                        level.addFreshEntity(endermite, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.ENDER_PEARL); // Paper - add reason
+                                    }
+                                }
+                                if (this.isOnPortalCooldown()) {
+                                    owner.setPortalCooldown();
+                                }
+                                if (newOwner != null) {
+                                    newOwner.resetFallDistance();
+                                    newOwner.resetCurrentImpulseContext();
+                                    newOwner.hurtServer(player.level(), this.damageSources().enderPearl().eventEntityDamager(this), 5.0F); // CraftBukkit // Paper - fix DamageSource API
+                                }
+                                this.playSound(level, teleportPos);
+                            });
+                        } else { // Fish end - Handle Enderpearl Teleport
                         ServerPlayer newOwner = player.teleport(new TeleportTransition(
                             level, teleportPos, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING).withCause(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (newOwner == null) {
@@ -141,6 +172,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         }
 
                         this.playSound(level, teleportPos);
+                        } // Fish - Handle Enderpearl Teleport
                     }
                 } else {
                     Entity newOwner = owner.teleport(
